sdk/python/kfp/dsl/executor.py

```diff
@@ -189,8 +201,29 @@ def handle_single_return_value(self, output_name: str, annotation_type: Any,
                     f'Function `{self.func.__name__}` returned value of type {type(return_value)}; want type {origin_type}'
                 )
             self.write_output_parameter_value(output_name, return_value)
+
         elif is_artifact(annotation_type):
-            self.write_output_artifact_payload(output_name, return_value)
+            if isinstance(return_value, artifact_types.Artifact):
+                # for -> Artifact annotations, where the user returns an artifact
+                artifact_name = self.executor_input['outputs']['artifacts'][
+                    output_name]['artifacts'][0]['name']
+                # users should not override the name for Vertex Pipelines
+                # if empty string, replace
+                # else provide descriptive warning and prefer letting backend throw exception
+                running_on_vertex = 'VERTEX_AI_PIPELINES_RUN_LABELS' in os.environ
+                if running_on_vertex:
+                    if return_value.name == '':
+                        return_value.name = artifact_name
+                    else:
+                        # prefer letting the backend throw the runtime exception
```

Comment:
Does Vertex Pipelines backend actually throw an error today? I was under the impression that every artifact property is mutable.

Reply:
The `name` field must conform to the Vertex API resourceName structure and it throws an error if it does not.

The error looks like this:

```
com.google.apps.framework.request.BadRequestException: [ResourceName error] Path 'foo' does not match template 'projects/{project_id_or_number=*}/locations/{location_id=*}/metadataStores/{metadata_store_id=*}/artifacts/{artifact_id=*}'.
```